dicoogle/src/main/java/pt/ua/dicoogle/taskManager/RunningIndexTasks.java

```diff
@@ -45,22 +47,24 @@
 	public static RunningIndexTasks instance;
 
 	private final Map<String, Task<Report>> taskRunningList;
+	private AtomicBoolean cleaning = new AtomicBoolean(false); // if cleaning task is running or not.
 
 	public static RunningIndexTasks getInstance() {
 		if (instance == null)
 			instance = new RunningIndexTasks();
-
 		return instance;
 	}
 
 	public RunningIndexTasks() {
-		taskRunningList = new LinkedHashMap<>();
+		taskRunningList = new ConcurrentHashMap<>(SOFT_MAX_RUNNINGTASKS, 0.75f, 4);
 	}
 
 	public void addTask(String taskUid, Task<Report> task) {
 		taskRunningList.put(taskUid, task);
-		if (ENABLE_HOOK){
+		if (ENABLE_HOOK && !cleaning.compareAndSet(false, true)){
+		    // will execute cleaning process
 		    hookRemoveRunningTasks();
+            cleaning.set(false); // already cleaned
```

Comment:
There is mixed whitespace in surrounding lines.
```suggestion
		    cleaning.set(false); // already cleaned
```

This should be fixed in Dicoogle 3, but we might as well keep it consistent here.

Reply:
Dicoogle3: To be honest, not sure if we should keep it consistent. We are just adding here some complexity that in the end we did not take advtange of tracking that list in the way we do. We can discuss it in another scope / hackathon.